postgres: SourceMetadata whole-DB scan races concurrent DDL (flaky TestDBSemver/Matview; sq inspect robustness)

## Summary

sq's Postgres `SourceMetadata` performs a whole-database scan (enumerate all
tables, then build per-table `pg_total_relation_size(...)` clauses). When tables
are created/dropped concurrently, the scan can:

1. build malformed SQL with an empty identifier, or
2. hit `could not open relation with OID ...`.

The full-version DB sweep (`go test ./...`, `scope=full`) runs multiple test
packages concurrently against a single Postgres database, each creating and
dropping tables, so the scan races their DDL. This is what #1024 observed as the
"possible flake" `TestDBSemver` / `TestPostgres_Matview` failures.

## Evidence

Running the full `drivers/postgres` suite against pg 9.6 / 10:

```
--- FAIL: TestDBSemver
    ERROR: syntax error at or near "") AS table_row_count,
      pg_total_relation_size('"" (SQLSTATE 42601)
--- FAIL: TestPostgres_Matview
    ERROR: could not open relation with OID 17164 (SQLSTATE XX000)
```

Both pass **3/3 in isolation** (`-run TestDBSemver` / `-run TestPostgres_Matview`),
so this is a concurrency artifact, not a version bug. It surfaces more on the
slower old-pg containers, which is why #1024 read it as a per-version flake.

## Root cause

`SourceMetadata` enumerates tables and then queries per-table sizes. A table
dropped between enumeration and the size query yields an empty name, so sq emits
`pg_total_relation_size('""...')` (invalid), or Postgres fails the query because
the relation's OID no longer resolves. sq does not tolerate a relation
disappearing mid-scan.

## Impact

- **Test flakiness** on the DB-integration sweep (sporadic red legs).
- **Real-world robustness**: `sq inspect` against a live database undergoing DDL
  can hit the same failure.

## Suggested fixes

- Make `SourceMetadata` resilient to relations vanishing mid-scan: never emit an
  empty/uninterpolated identifier; skip or tolerate missing relations; consider a
  single catalog-consistent snapshot query.
- For the tests specifically, isolate per-package test data (separate schema) to
  remove the cross-package DDL race.

## Priority

Medium — sporadic CI red plus a latent `sq inspect` robustness gap. Surfaced
while fixing the #1024 legacy-Postgres cluster; orthogonal to that version work.

--- a/.github/workflows/db-integration.yml
+++ b/.github/workflows/db-integration.yml
@@ -120,6 +120,27 @@ jobs:
         with:
           go-version-file: go.mod
 
+      # scope=full runs `go test ./...`, whose embedded SQLite/DuckDB temp
+      # databases can exhaust the runner's ~14 GB root filesystem (each leg runs
+      # the full suite on its own runner, so this bites single-version nightlies
+      # too, not just multi-version dispatches). Reclaim space from preinstalled
+      # toolchains we never use. Not needed for scope=narrow.
+      - name: Free disk space
+        if: ${{ (inputs.scope || github.event.inputs.scope || 'full') == 'full' }}
+        shell: bash
+        run: |
+          set -euo pipefail
+          before=$(df -BG --output=avail / | tail -1 | tr -dc '0-9')
+          sudo rm -rf \
+            /usr/local/lib/android \
+            /opt/hostedtoolcache/CodeQL \
+            /usr/share/dotnet \
+            /opt/ghc \
+            /usr/local/.ghcup \
+            /usr/share/swift
+          after=$(df -BG --output=avail / | tail -1 | tr -dc '0-9')
+          echo "Freed disk on /: ${before}G -> ${after}G available"
+
       - name: Wait for DB healthy
         shell: bash
         run: |

--- a/.github/workflows/db-integration.yml
+++ b/.github/workflows/db-integration.yml
@@ -120,11 +120,20 @@ jobs:
         with:
           go-version-file: go.mod
 
-      # scope=full runs `go test ./...`, whose embedded SQLite/DuckDB temp
-      # databases can exhaust the runner's ~14 GB root filesystem (each leg runs
-      # the full suite on its own runner, so this bites single-version nightlies
-      # too, not just multi-version dispatches). Reclaim space from preinstalled
-      # toolchains we never use. Not needed for scope=narrow.
+      # scope=full runs `go test ./...`, whose many embedded SQLite/DuckDB temp
+      # databases exhaust the runner (observed: a pg17 leg hit "no space left on
+      # device"). The runner has ~86 GB free on / (measured: 86G -> 110G after
+      # this step), so it is almost certainly inode exhaustion, not bytes: the
+      # preinstalled toolchains removed below are millions of small files, so the
+      # rm frees inodes as much as the ~20 GB. Each leg runs the full suite on
+      # its own runner, so this bites single-version nightlies too, not just
+      # multi-version dispatches.
+      #
+      # TODO: gated to scope=full because narrow runs only the per-engine
+      # packages and has not exhausted in practice. That is an empirical
+      # assumption, not a guarantee — if narrow's libsq/cli cross-source tests
+      # grow heavier it may need this too, at which point drop the `if:` and
+      # always free space.
       - name: Free disk space
         if: ${{ (inputs.scope || github.event.inputs.scope || 'full') == 'full' }}
         shell: bash

--- a/drivers/postgres/errors.go
+++ b/drivers/postgres/errors.go
@@ -2,7 +2,6 @@ package postgres
 
 import (
 	"errors"
-	"strings"
 
 	"github.com/jackc/pgx/v5/pgconn"
 
@@ -50,22 +49,17 @@ func isErrRelationNotExist(err error) bool {
 	return hasErrCode(err, errCodeRelationNotExist)
 }
 
-// isErrRelationDroppedMidScan reports whether err indicates a relation
-// disappeared while a source-wide metadata scan was reading it: the canonical
-// "relation does not exist" (42P01), or the lower-level "could not open
-// relation with OID ..." (XX000) that pg_total_relation_size / regclass raise
-// when a relation is dropped between OID resolution and access. A scan reads a
-// live database, so it tolerates such a drop rather than failing.
-func isErrRelationDroppedMidScan(err error) bool {
-	if isErrRelationNotExist(err) {
-		return true
-	}
-	var pgErr *pgconn.PgError
-	if errors.As(err, &pgErr) {
-		return pgErr.Code == errCodeInternalError &&
-			strings.Contains(pgErr.Message, "could not open relation")
-	}
-	return false
+// isErrScanRetryable reports whether a bulk metadata-loader error is worth
+// retrying during a source-wide scan of a live database: too-many-connections,
+// or a relation that vanished mid-query. The latter surfaces either as 42P01,
+// or as an XX000 internal error (e.g. "could not open relation with OID") when a
+// relation is dropped between resolution and access. Matching XX000 by code
+// (not message text) keeps this locale-independent, and retrying is safe: a
+// persistent error still surfaces once the retry budget is exhausted.
+func isErrScanRetryable(err error) bool {
+	return isErrTooManyConnections(err) ||
+		isErrRelationNotExist(err) ||
+		hasErrCode(err, errCodeInternalError)
 }
 
 // hasErrCode returns true if err (or its cause error)

--- a/drivers/postgres/errors.go
+++ b/drivers/postgres/errors.go
@@ -2,6 +2,7 @@ package postgres
 
 import (
 	"errors"
+	"strings"
 
 	"github.com/jackc/pgx/v5/pgconn"
 
@@ -27,6 +28,10 @@ func errw(err error) error {
 const (
 	errCodeRelationNotExist   = "42P01"
 	errCodeTooManyConnections = "53300"
+	// errCodeInternalError (XX000) is raised, among other cases, as "could not
+	// open relation with OID ..." when a relation is dropped between OID
+	// resolution and access, i.e. by concurrent DDL.
+	errCodeInternalError = "XX000"
 )
 
 // isErrTooManyConnections returns true if err is a postgres error
@@ -43,6 +48,24 @@ func isErrTooManyConnections(err error) bool {
 // See: https://www.postgresql.org/docs/14/errcodes-appendix.html
 func isErrRelationNotExist(err error) bool {
 	return hasErrCode(err, errCodeRelationNotExist)
+}
+
+// isErrRelationDroppedMidScan reports whether err indicates a relation
+// disappeared while a source-wide metadata scan was reading it: the canonical
+// "relation does not exist" (42P01), or the lower-level "could not open
+// relation with OID ..." (XX000) that pg_total_relation_size / regclass raise
+// when a relation is dropped between OID resolution and access. A scan reads a
+// live database, so it tolerates such a drop rather than failing.
+func isErrRelationDroppedMidScan(err error) bool {
+	if isErrRelationNotExist(err) {
+		return true
+	}
+	var pgErr *pgconn.PgError
+	if errors.As(err, &pgErr) {
+		return pgErr.Code == errCodeInternalError &&
+			strings.Contains(pgErr.Message, "could not open relation")
+	}
+	return false
 }
 
 // hasErrCode returns true if err (or its cause error)

--- a/drivers/postgres/metadata.go
+++ b/drivers/postgres/metadata.go
@@ -184,6 +184,23 @@ func toNullableScanType(log *slog.Logger, colName, dbTypeName string, knd kind.K
 	return nullableScanType
 }
 
+// relationExistsInCurrentSchema reports whether a relation named tblName exists
+// in the current schema. It distinguishes a table dropped mid-scan (tolerate)
+// from a genuine error on a table that still exists (propagate), independently
+// of server locale. The name is a bound parameter compared against pg_class,
+// so no identifier escaping is needed.
+func relationExistsInCurrentSchema(ctx context.Context, db sqlz.DB, tblName string) (bool, error) {
+	const q = `SELECT EXISTS(
+  SELECT 1 FROM pg_catalog.pg_class c
+  JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
+  WHERE c.relname = $1 AND n.nspname = current_schema())`
+	var exists bool
+	if err := db.QueryRowContext(ctx, q, tblName).Scan(&exists); err != nil {
+		return false, errw(err)
+	}
+	return exists, nil
+}
+
 // loadWithRetry runs a bulk metadata loader with vanished-relation retry, so a
 // table dropped mid-scan by concurrent DDL doesn't fail the whole source scan.
 func loadWithRetry[T any](ctx context.Context, load func() (T, error)) (T, error) {
@@ -272,19 +289,20 @@ current_setting('server_version'), version(), "current_user"()`
 			})
 
 			if mdErr != nil {
-				switch {
-				case isErrRelationDroppedMidScan(mdErr):
-					// If the table is dropped while we're collecting metadata
-					// (42P01, or the lower-level "could not open relation with
-					// OID"), log a warning and suppress the error.
-					log.Warn(
-						"metadata collection: table not found (continuing regardless)",
-						lga.Table, tblNames[i],
-						lga.Err, mdErr,
-					)
-				default:
+				// A relation dropped by concurrent DDL mid-scan is expected on a
+				// live database. Confirm the table is actually gone (a
+				// locale-independent check that, unlike matching the error text,
+				// won't mask a genuine fault on a table that still exists), then
+				// tolerate it; otherwise the error is real and propagates.
+				exists, existErr := relationExistsInCurrentSchema(gCtx, db, tblNames[i])
+				if existErr != nil || exists {
 					return mdErr
 				}
+				log.Warn(
+					"metadata collection: table dropped during scan (continuing regardless)",
+					lga.Table, tblNames[i],
+					lga.Err, mdErr,
+				)
 			}
 
 			tblMetas[i] = tblMeta
@@ -554,10 +572,10 @@ AND t.table_name = $1`
 // pg_catalog. Matviews are absent from information_schema, so this is the
 // fallback path from getTableMetadata. It is a two-step operation: first
 // confirm that name actually is a matview (returning the canonical not-found
-// error otherwise), and only then run the detail query. The size/comment/
-// viewdef functions take name as a bound quote_ident($1)::regclass parameter; only the
-// COUNT(*) FROM identifier is interpolated, and only after Step A confirms
-// name is a real matview.
+// error otherwise), and only then run the detail query. The size/comment/viewdef
+// functions take name via quote_ident($1)::regclass; only the COUNT(*) FROM
+// identifier is interpolated, and only after Step A confirms name is a real
+// matview.
 func getMatviewMetadata(ctx context.Context, db sqlz.DB, name string) (*metadata.Table, error) {
 	// Step A: confirm name is a matview in the current schema, and get
 	// the catalog & schema for the FQ name.
@@ -581,7 +599,7 @@ WHERE EXISTS (
 	debugz.DebugSleep(ctx)
 
 	// Step B: name is confirmed a matview. The size/comment/viewdef
-	// functions take name as a bound quote_ident($1)::regclass parameter. Only the
+	// functions take name via quote_ident($1)::regclass. Only the
 	// COUNT(*) FROM identifier must be interpolated: a relation name
 	// cannot be a bind parameter in a FROM clause, and it's safe here
 	// because Step A confirmed name is a real matview in this schema.

--- a/drivers/postgres/metadata.go
+++ b/drivers/postgres/metadata.go
@@ -184,6 +184,18 @@ func toNullableScanType(log *slog.Logger, colName, dbTypeName string, knd kind.K
 	return nullableScanType
 }
 
+// loadWithRetry runs a bulk metadata loader with vanished-relation retry, so a
+// table dropped mid-scan by concurrent DDL doesn't fail the whole source scan.
+func loadWithRetry[T any](ctx context.Context, load func() (T, error)) (T, error) {
+	var result T
+	err := doRetryVanished(ctx, func() error {
+		var e error
+		result, e = load()
+		return e
+	})
+	return result, err
+}
+
 func getSourceMetadata(ctx context.Context, src *source.Source, db sqlz.DB, noSchema bool) (*metadata.Source, error) {
 	log := lg.FromContext(ctx)
 	ctx = options.NewContext(ctx, src.Options)
@@ -261,9 +273,10 @@ current_setting('server_version'), version(), "current_user"()`
 
 			if mdErr != nil {
 				switch {
-				case isErrRelationNotExist(mdErr):
-					// For example, if the table is dropped while we're collecting
-					// metadata, we log a warning and suppress the error.
+				case isErrRelationDroppedMidScan(mdErr):
+					// If the table is dropped while we're collecting metadata
+					// (42P01, or the lower-level "could not open relation with
+					// OID"), log a warning and suppress the error.
 					log.Warn(
 						"metadata collection: table not found (continuing regardless)",
 						lga.Table, tblNames[i],
@@ -296,39 +309,54 @@ current_setting('server_version'), version(), "current_user"()`
 
 	md.RecomputeTableCounts()
 
-	// Fetch foreign keys for all tables in one query, assign them to
-	// their owning tables, and derive the cross-table back-references.
-	allFKs, err := getPgForeignKeys(ctx, db, "")
+	// Fetch foreign keys for all tables in one query, assign them to their
+	// owning tables, and derive the cross-table back-references. Each bulk
+	// loader is a single catalog query over every table in the schema; a table
+	// dropped mid-query by concurrent DDL is retried (loadWithRetry) so the
+	// scan doesn't fail on transient churn.
+	allFKs, err := loadWithRetry(ctx, func() ([]*metadata.ForeignKey, error) {
+		return getPgForeignKeys(ctx, db, "")
+	})
 	if err != nil {
 		return nil, err
 	}
 	metadata.AssignForeignKeys(log, md.Tables, allFKs)
 
-	allUCs, err := getPgUniqueConstraints(ctx, db, "")
+	allUCs, err := loadWithRetry(ctx, func() ([]*metadata.UniqueConstraint, error) {
+		return getPgUniqueConstraints(ctx, db, "")
+	})
 	if err != nil {
 		return nil, err
 	}
 	metadata.AssignUniqueConstraints(log, md.Tables, allUCs)
 
-	allChecks, err := getPgCheckConstraints(ctx, db, "")
+	allChecks, err := loadWithRetry(ctx, func() ([]*metadata.CheckConstraint, error) {
+		return getPgCheckConstraints(ctx, db, "")
+	})
 	if err != nil {
 		return nil, err
 	}
 	metadata.AssignCheckConstraints(log, md.Tables, allChecks)
 
-	allTriggers, err := getPgTriggers(ctx, db, "")
+	allTriggers, err := loadWithRetry(ctx, func() ([]*metadata.Trigger, error) {
+		return getPgTriggers(ctx, db, "")
+	})
 	if err != nil {
 		return nil, err
 	}
 	metadata.AssignTriggers(log, md.Tables, allTriggers)
 
-	allIdxs, err := getPgIndexes(ctx, db, "")
+	allIdxs, err := loadWithRetry(ctx, func() ([]*metadata.Index, error) {
+		return getPgIndexes(ctx, db, "")
+	})
 	if err != nil {
 		return nil, err
 	}
 	metadata.AssignIndexes(log, md.Tables, allIdxs)
 
-	allViewDefs, err := getPgViewDefinitions(ctx, db, "")
+	allViewDefs, err := loadWithRetry(ctx, func() (map[string]string, error) {
+		return getPgViewDefinitions(ctx, db, "")
+	})
 	if err != nil {
 		return nil, err
 	}
@@ -448,16 +476,25 @@ ORDER BY table_name`
 func getTableMetadata(ctx context.Context, db sqlz.DB, tblName string) (*metadata.Table, error) {
 	log := lg.FromContext(ctx)
 
-	const tblsQueryTpl = `SELECT table_catalog, table_schema, table_name, table_type, is_insertable_into,
-  (SELECT COUNT(*) FROM "%s") AS table_row_count,
-  pg_total_relation_size('%q') AS table_size,
-  (SELECT '%q'::regclass::oid AS table_oid),
-  obj_description('%q'::REGCLASS, 'pg_class') AS table_comment
-FROM information_schema.tables
-WHERE table_catalog = current_database()
-AND table_schema = current_schema()
-AND table_name = $1`
-	tablesQuery := fmt.Sprintf(tblsQueryTpl, tblName, tblName, tblName, tblName)
+	// A table name can legally contain a double-quote (e.g. `we"ird`). The
+	// relation's OID is resolved via a pg_class JOIN on the raw name ($1, a
+	// bind parameter) rather than a regclass text cast, which mis-parses such
+	// names; the size/oid/comment functions then take the OID directly. The
+	// only interpolation is the COUNT subquery's FROM, which uses a
+	// double-quoted (escaped) identifier. Raw interpolation built malformed
+	// SQL. See issue #1025.
+	safeIdent := stringz.DoubleQuote(tblName)
+	tablesQuery := `SELECT t.table_catalog, t.table_schema, t.table_name, t.table_type, t.is_insertable_into,
+  (SELECT COUNT(*) FROM ` + safeIdent + `) AS table_row_count,
+  pg_total_relation_size(c.oid) AS table_size,
+  c.oid AS table_oid,
+  obj_description(c.oid, 'pg_class') AS table_comment
+FROM information_schema.tables t
+JOIN pg_catalog.pg_class c ON c.relname = t.table_name
+JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace AND n.nspname = t.table_schema
+WHERE t.table_catalog = current_database()
+AND t.table_schema = current_schema()
+AND t.table_name = $1`
 
 	pgTbl := &pgTable{}
 	err := db.QueryRowContext(ctx, tablesQuery, tblName).
@@ -518,7 +555,7 @@ AND table_name = $1`
 // fallback path from getTableMetadata. It is a two-step operation: first
 // confirm that name actually is a matview (returning the canonical not-found
 // error otherwise), and only then run the detail query. The size/comment/
-// viewdef functions take name as a bound $1::regclass parameter; only the
+// viewdef functions take name as a bound quote_ident($1)::regclass parameter; only the
 // COUNT(*) FROM identifier is interpolated, and only after Step A confirms
 // name is a real matview.
 func getMatviewMetadata(ctx context.Context, db sqlz.DB, name string) (*metadata.Table, error) {
@@ -544,15 +581,15 @@ WHERE EXISTS (
 	debugz.DebugSleep(ctx)
 
 	// Step B: name is confirmed a matview. The size/comment/viewdef
-	// functions take name as a bound $1::regclass parameter. Only the
+	// functions take name as a bound quote_ident($1)::regclass parameter. Only the
 	// COUNT(*) FROM identifier must be interpolated: a relation name
 	// cannot be a bind parameter in a FROM clause, and it's safe here
 	// because Step A confirmed name is a real matview in this schema.
 	const detailQueryTpl = `SELECT
   (SELECT COUNT(*) FROM "%s") AS row_count,
-  pg_total_relation_size($1::regclass) AS mv_size,
-  obj_description($1::regclass, 'pg_class') AS mv_comment,
-  pg_get_viewdef($1::regclass, true) AS view_def`
+  pg_total_relation_size(quote_ident($1)::regclass) AS mv_size,
+  obj_description(quote_ident($1)::regclass, 'pg_class') AS mv_comment,
+  pg_get_viewdef(quote_ident($1)::regclass, true) AS view_def`
 	// Double any embedded double-quotes so the identifier literal can't be broken out of.
 	safeName := strings.ReplaceAll(name, `"`, `""`)
 	detailQuery := fmt.Sprintf(detailQueryTpl, safeName)
@@ -823,7 +860,7 @@ func getPgColumns(ctx context.Context, db sqlz.DB, tblName string) ([]*pgColumn,
 	FROM
 		pg_catalog.pg_class c
 	WHERE
-		c.oid = (SELECT ('"' || cols.table_name || '"')::regclass::oid)
+		c.oid = (SELECT quote_ident(cols.table_name)::regclass::oid)
 		AND c.relname = cols.table_name
 	) AS column_comment
 FROM information_schema.columns cols
@@ -902,14 +939,14 @@ func getPgConstraints(ctx context.Context, db sqlz.DB, tblName string) ([]*pgCon
     (
        SELECT pg_catalog.pg_get_constraintdef(pgc.oid, TRUE)
        FROM pg_catalog.pg_constraint pgc
-       WHERE pgc.conrelid = (SELECT ('"' || kcu.table_name || '"')::regclass::oid)
+       WHERE pgc.conrelid = (SELECT quote_ident(kcu.table_name)::regclass::oid)
        AND pgc.conname = tc.constraint_name
        limit 1
     )  AS constraint_def,
     (
        SELECT pgc.confrelid::regclass
        FROM pg_catalog.pg_constraint pgc
-       WHERE pgc.conrelid = (SELECT ('"' || kcu.table_name || '"')::regclass::oid)
+       WHERE pgc.conrelid = (SELECT quote_ident(kcu.table_name)::regclass::oid)
        AND pgc.conname = tc.constraint_name
        AND pgc.confrelid > 0
        LIMIT 1

--- a/drivers/postgres/metadata.go
+++ b/drivers/postgres/metadata.go
@@ -184,23 +184,6 @@ func toNullableScanType(log *slog.Logger, colName, dbTypeName string, knd kind.K
 	return nullableScanType
 }
 
-// relationExistsInCurrentSchema reports whether a relation named tblName exists
-// in the current schema. It distinguishes a table dropped mid-scan (tolerate)
-// from a genuine error on a table that still exists (propagate), independently
-// of server locale. The name is a bound parameter compared against pg_class,
-// so no identifier escaping is needed.
-func relationExistsInCurrentSchema(ctx context.Context, db sqlz.DB, tblName string) (bool, error) {
-	const q = `SELECT EXISTS(
-  SELECT 1 FROM pg_catalog.pg_class c
-  JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
-  WHERE c.relname = $1 AND n.nspname = current_schema())`
-	var exists bool
-	if err := db.QueryRowContext(ctx, q, tblName).Scan(&exists); err != nil {
-		return false, errw(err)
-	}
-	return exists, nil
-}
-
 // loadWithRetry runs a bulk metadata loader with vanished-relation retry, so a
 // table dropped mid-scan by concurrent DDL doesn't fail the whole source scan.
 func loadWithRetry[T any](ctx context.Context, load func() (T, error)) (T, error) {
@@ -289,20 +272,18 @@ current_setting('server_version'), version(), "current_user"()`
 			})
 
 			if mdErr != nil {
-				// A relation dropped by concurrent DDL mid-scan is expected on a
-				// live database. Confirm the table is actually gone (a
-				// locale-independent check that, unlike matching the error text,
-				// won't mask a genuine fault on a table that still exists), then
-				// tolerate it; otherwise the error is real and propagates.
-				exists, existErr := relationExistsInCurrentSchema(gCtx, db, tblNames[i])
-				if existErr != nil || exists {
+				switch {
+				case isErrRelationNotExist(mdErr):
+					// For example, if the table is dropped while we're collecting
+					// metadata, we log a warning and suppress the error.
+					log.Warn(
+						"metadata collection: table not found (continuing regardless)",
+						lga.Table, tblNames[i],
+						lga.Err, mdErr,
+					)
+				default:
 					return mdErr
 				}
-				log.Warn(
-					"metadata collection: table dropped during scan (continuing regardless)",
-					lga.Table, tblNames[i],
-					lga.Err, mdErr,
-				)
 			}
 
 			tblMetas[i] = tblMeta

--- a/drivers/postgres/metadata_test.go
+++ b/drivers/postgres/metadata_test.go
@@ -291,10 +291,11 @@ func TestPostgres_ColumnFlags(t *testing.T) {
 }
 
 // TestPostgres_QuotedTableName_Metadata pins that table metadata loads for a
-// table whose name contains a double-quote. getTableMetadata must escape the
-// identifier (and pass it as a bound regclass parameter); raw interpolation
-// built malformed SQL, which surfaced as a flake when a full-source scan
-// happened to run while such a table existed (issue #1025).
+// table whose name contains a double-quote. getTableMetadata must resolve the
+// OID via a pg_class JOIN on the bound name and use an escaped identifier in the
+// COUNT subquery; raw interpolation built malformed SQL, which surfaced as a
+// flake when a full-source scan happened to run while such a table existed
+// (issue #1025).
 func TestPostgres_QuotedTableName_Metadata(t *testing.T) {
 	tu.SkipShort(t, true)
 	t.Parallel()

--- a/drivers/postgres/metadata_test.go
+++ b/drivers/postgres/metadata_test.go
@@ -290,6 +290,35 @@ func TestPostgres_ColumnFlags(t *testing.T) {
 	}
 }
 
+// TestPostgres_QuotedTableName_Metadata pins that table metadata loads for a
+// table whose name contains a double-quote. getTableMetadata must escape the
+// identifier (and pass it as a bound regclass parameter); raw interpolation
+// built malformed SQL, which surfaced as a flake when a full-source scan
+// happened to run while such a table existed (issue #1025).
+func TestPostgres_QuotedTableName_Metadata(t *testing.T) {
+	tu.SkipShort(t, true)
+	t.Parallel()
+
+	th := testh.New(t)
+	src := th.Source(sakila.Pg)
+	db := th.OpenDB(src)
+
+	tbl := stringz.UniqTableName(`me"ta`)
+	qtbl := stringz.DoubleQuote(tbl)
+	_, err := db.ExecContext(th.Context, `CREATE TABLE `+qtbl+` (id INT, name TEXT)`)
+	require.NoError(t, err)
+	t.Cleanup(func() { th.DropTable(src, tablefq.From(tbl)) })
+
+	_, err = db.ExecContext(th.Context, `INSERT INTO `+qtbl+` (id, name) VALUES (1, 'a'), (2, 'b')`)
+	require.NoError(t, err)
+
+	md, err := th.Open(src).TableMetadata(th.Context, tbl)
+	require.NoError(t, err)
+	require.Equal(t, tbl, md.Name)
+	require.Equal(t, int64(2), md.RowCount, "RowCount must load for a quoted table name")
+	require.Len(t, md.Columns, 2)
+}
+
 // TestPostgres_ViewDefinition verifies that ViewDefinition is populated for
 // views and is empty for base tables.
 func TestPostgres_ViewDefinition(t *testing.T) {

--- a/drivers/postgres/metadata_test.go
+++ b/drivers/postgres/metadata_test.go
@@ -68,6 +68,12 @@ func TestIndexes_IncludeFilter_Postgres(t *testing.T) {
 	src := th.Source(sakila.Pg)
 	db := th.OpenDB(src)
 
+	// INCLUDE covering indexes were added in Postgres 11; there is no pre-11
+	// equivalent to introspect.
+	if !th.DBSemverAtLeast(sakila.Pg, "v11.0.0") {
+		t.Skip("INCLUDE covering indexes require Postgres >= 11")
+	}
+
 	tbl := stringz.UniqTableName("idx_include")
 	_, err := db.ExecContext(th.Context,
 		"CREATE TABLE "+tbl+" (k INT, extra TEXT)")
@@ -235,34 +241,53 @@ func TestPostgres_ColumnFlags(t *testing.T) {
 	src := th.Source(sakila.Pg)
 	db := th.OpenDB(src)
 
+	// Column flags are version-gated: identity columns arrived in Postgres 10,
+	// stored generated columns in Postgres 12. Collation applies on every
+	// supported version, so the base table always exercises it.
+	hasIdentity := th.DBSemverAtLeast(sakila.Pg, "v10.0.0")
+	hasGenerated := th.DBSemverAtLeast(sakila.Pg, "v12.0.0")
+
 	tbl := stringz.UniqTableName("col_flags")
 	_, err := db.ExecContext(th.Context,
 		`CREATE TABLE `+tbl+` (
-			id         bigint GENERATED ALWAYS AS IDENTITY,
 			first_name text,
 			last_name  text,
-			full_name  text GENERATED ALWAYS AS (first_name || ' ' || last_name) STORED,
 			country    text COLLATE "C"
 		)`)
 	require.NoError(t, err)
 	t.Cleanup(func() { th.DropTable(src, tablefq.From(tbl)) })
 
+	if hasIdentity {
+		_, err = db.ExecContext(th.Context,
+			`ALTER TABLE `+tbl+` ADD COLUMN id bigint GENERATED ALWAYS AS IDENTITY`)
+		require.NoError(t, err)
+	}
+	if hasGenerated {
+		_, err = db.ExecContext(th.Context,
+			`ALTER TABLE `+tbl+` ADD COLUMN full_name text `+
+				`GENERATED ALWAYS AS (first_name || ' ' || last_name) STORED`)
+		require.NoError(t, err)
+	}
+
 	md, err := th.Open(src).TableMetadata(th.Context, tbl)
 	require.NoError(t, err)
-
-	idCol := md.Column("id")
-	require.NotNil(t, idCol)
-	require.True(t, idCol.Identity, "id: Identity should be true")
-
-	fullNameCol := md.Column("full_name")
-	require.NotNil(t, fullNameCol)
-	require.True(t, fullNameCol.Generated, "full_name: Generated should be true")
-	require.Contains(t, fullNameCol.GeneratedExpr, "first_name",
-		"full_name: GeneratedExpr should contain 'first_name'")
 
 	countryCol := md.Column("country")
 	require.NotNil(t, countryCol)
 	require.Equal(t, "C", countryCol.Collation, "country: Collation should be 'C'")
+
+	if hasIdentity {
+		idCol := md.Column("id")
+		require.NotNil(t, idCol)
+		require.True(t, idCol.Identity, "id: Identity should be true")
+	}
+	if hasGenerated {
+		fullNameCol := md.Column("full_name")
+		require.NotNil(t, fullNameCol)
+		require.True(t, fullNameCol.Generated, "full_name: Generated should be true")
+		require.Contains(t, fullNameCol.GeneratedExpr, "first_name",
+			"full_name: GeneratedExpr should contain 'first_name'")
+	}
 }
 
 // TestPostgres_ViewDefinition verifies that ViewDefinition is populated for
@@ -331,7 +356,7 @@ func TestPostgres_Triggers(t *testing.T) {
 
 	_, err = db.ExecContext(th.Context,
 		`CREATE TRIGGER `+trigName+` AFTER INSERT OR UPDATE ON `+tbl+
-			` FOR EACH ROW EXECUTE FUNCTION `+fnName+`()`)
+			` FOR EACH ROW EXECUTE PROCEDURE `+fnName+`()`)
 	require.NoError(t, err)
 
 	md, err := th.Open(src).TableMetadata(th.Context, tbl)
@@ -393,7 +418,7 @@ func TestPostgres_ViewInsteadOfTrigger(t *testing.T) {
 
 	_, err = db.ExecContext(th.Context,
 		`CREATE TRIGGER `+trigName+` INSTEAD OF INSERT ON `+viewName+
-			` FOR EACH ROW EXECUTE FUNCTION `+fnName+`()`)
+			` FOR EACH ROW EXECUTE PROCEDURE `+fnName+`()`)
 	require.NoError(t, err)
 	// Dropping the view removes its triggers; no separate trigger DROP needed.
 

--- a/drivers/postgres/postgres.go
+++ b/drivers/postgres/postgres.go
@@ -900,3 +900,13 @@ func doRetry(ctx context.Context, fn func() error) error {
 	maxRetryInterval := tuning.OptMaxRetryInterval.Get(options.FromContext(ctx))
 	return retry.Do(ctx, maxRetryInterval, fn, isErrTooManyConnections)
 }
+
+// doRetryVanished is like doRetry, but also retries when a relation vanishes
+// mid-operation because of concurrent DDL. A source-wide metadata scan issues
+// bulk catalog queries against a live database, so a table dropped mid-query is
+// transient; retrying lets the churn settle. The error is still returned if
+// retries are exhausted.
+func doRetryVanished(ctx context.Context, fn func() error) error {
+	maxRetryInterval := tuning.OptMaxRetryInterval.Get(options.FromContext(ctx))
+	return retry.Do(ctx, maxRetryInterval, fn, isErrTooManyConnections, isErrRelationDroppedMidScan)
+}

--- a/drivers/postgres/postgres.go
+++ b/drivers/postgres/postgres.go
@@ -908,5 +908,5 @@ func doRetry(ctx context.Context, fn func() error) error {
 // retries are exhausted.
 func doRetryVanished(ctx context.Context, fn func() error) error {
 	maxRetryInterval := tuning.OptMaxRetryInterval.Get(options.FromContext(ctx))
-	return retry.Do(ctx, maxRetryInterval, fn, isErrTooManyConnections, isErrRelationDroppedMidScan)
+	return retry.Do(ctx, maxRetryInterval, fn, isErrScanRetryable)
 }

